### ISSUE TITLE
switch to glib logging framework

### ISSUE THREAD

**Issue by puzrin** [jcupitt/libvips#544]
https://github.com/lovell/sharp/issues/607

libvips can write warning to console in some cases:

```
vips warning: VipsJpeg: bad exif meta "exif-ifd3-GPSVersionID"
vips warning: icc_transform: embedded profile incompatible with image
```

Such behaviour is unexpected and annoying. For example, it can break progress bars in terminal. Is is possible to disable warnings by default (or via config option)? Disabling via env looks like kludge IMHO. I'd like libvips "work right out of box".

**Comment by jcupitt** [comment]
The warnings are printed to stderr, so another option is to send that to /dev/null. Of course that would also hide all error messages. 

Warnings are supposed to be about serious things and should not be hidden by default. I agree that trivial warnings are not helpful. 

glib has a logging framework which libvips should probably use:

https://developer.gnome.org/glib/stable/glib-Message-Logging.html

That would give a uniform set of controls over message importance, where messages go, and what the response should be. Would that be a useful change?

**Comment by puzrin** [comment]
I think, everything depends on context. If you create app with GUI, then stdout/stderr content is not principal. But if you create CLI/server app, many of those wish stdout/stderr to be controlled and predictable. For example, to write logs in JSON format.

Personally, i'm ok with any change that stops writing to stdout/stderr. Also, i'll be ok if `sharp` will be able to override such defaults (without env change).

I'm not lib writer and can't suggest concrete method. Just describe situation from user's point of view.

**Comment by jcupitt** [comment]
OK, let's put switching to glib logging on the TODO list for 8.5. I'll update the title to reflect this.

**Comment by jcupitt** [comment]
Made a branch for this improvement. 

https://github.com/jcupitt/libvips/tree/glib-logging

**Comment by jcupitt** [comment]
OK, merged and done!

`--vips-info` still works for getting info logging messages, but the GLib logging stuff is now the preferred way to do it. For example:

```
$ G_MESSAGES_DEBUG=VIPS vipsthumbnail k2.jpg
VIPS-INFO: thumbnailing k2.jpg
VIPS-INFO: selected loader is VipsForeignLoadJpegFile
VIPS-INFO: input size is 1450 x 2048
VIPS-INFO: loading jpeg with factor 8 pre-shrink
VIPS-INFO: converting to processing space srgb
VIPS-INFO: residual reducev by 0.5
VIPS-INFO: 13 point mask
VIPS-INFO: using vector path
VIPS-INFO: residual reduceh by 0.5
VIPS-INFO: 13 point mask
VIPS-INFO: thumbnailing k2.jpg as ./tn_k2.jpg
```

Warning messages now look like this:

```
$ vips copy x_truncated.jpg x.tif

(vips:519): VIPS-WARNING **: read gave 2 warnings

(vips:519): VIPS-WARNING **: VipsJpeg: Premature end of JPEG file
```

They are displayed if you run in a terminal, or added to the systemd log if you run as a service. You can intercept these messages by defining your own handler.

https://developer.gnome.org/glib/stable/glib-Message-Logging.html

**Comment by puzrin** [comment]
Great! Hope to see this in `sharp` soon. This change is very important.

**Comment by Nakilon** [comment]
Excuse me, how do I disable this logging at all?
```
   VIPS-INFO **:residual reducev by 0.430109 
.../gems/gobject-introspection-3.0.8/lib/gobject-introspection/loader.rb: line 551
   VIPS-INFO **:reducev by factor 2.32499
.../gems/gobject-introspection-3.0.8/lib/gobject-introspec
tion/loader.rb: line 551
   VIPS-INFO **:15 point mask
.../gems/gobject-introspection-3.0.8/lib/gobject-introspection/loader.rb: line 551
   VIPS-INFO **:reducev sequential line cache
.../gems/gobject-introspection-3.0.8/lib/gobject-introspection/loader.rb: line 551
   VIPS-INFO **:residual reduceh by 0.430109 
.../gems/gobject-introspection-3.0.8/lib/gobject-introspection/loader.rb: line 551
   VIPS-INFO **:15 point mask
```

**Comment by jcupitt** [comment]
It's off by default, perhaps something in your centos environment enabled it?

I would check for the `G_MESSAGES_DEBUG` and `VIPS_INFO` environment variables, and the `--vips-info` command-line flag.

**Comment by Nakilon** [comment]
```
declare -x CLICOLOR="1"
declare -x EDITOR="nano"
declare -x GEM_HOME="./gems21"
declare -x GI_TYPELIB_PATH="/usr/local/lib/girepository-1.0"
declare -x HISTSIZE="9001"
declare -x HOME="/home/myuser"
declare -x HOSTNAME="lab-2"
declare -x LANG="en_US.UTF-8"
declare -x LC_CTYPE="en_US.UTF-8"
declare -x LD_LIBRARY_PATH="/usr/local/lib"
declare -x LESSOPEN="||/usr/bin/lesspipe.sh %s"
declare -x LOGNAME="myuser"
declare -x LS_COLORS="....."
declare -x MAIL="/var/spool/mail/mysudoer"
declare -x OLDPWD="/home/mysudoer"
declare -x PATH="....."
declare -x PWD="/home/myuser/..."
declare -x RBENV_SHELL="bash"
declare -x RBENV_VERSION="2.1.10"
declare -x SHELL="/bin/bash"
declare -x SHLVL="1"
declare -x SUDO_COMMAND="/bin/su myuser"
declare -x SUDO_GID="1001"
declare -x SUDO_UID="1000"
declare -x SUDO_USER="mysudoer"
declare -x TERM="screen"
declare -x USER="myuser"
declare -x USERNAME="root"
declare -x XDG_SESSION_ID="3065"
```

**Comment by jcupitt** [comment]
I guess you don't see messages if you run command-line vips stuff, is that right? For example:

```
vipsthumbnail somefile.jpg
```

Is it just from Ruby?

**Comment by Nakilon** [comment]
`$ vipsthumbnail w_thumb.temp` printed nothing.

**Comment by jcupitt** [comment]
Could you try `irb` then? What do you see for:

```
$ irb
irb(main):001:0> require 'vips'
=> true
irb(main):002:0> x = Vips::Image.new_from_file "k2.jpg"
=> #<Vips::Image:0x55a3579ccd68 ptr=0x55a3580d5000>
irb(main):003:0> x.write_to_file "x.jpg"
=> nil
irb(main):004:0> 
```

**Comment by felixbuenemann** [comment]
This change appears to break the behavior of the `VIPS_WARNING` environment variable. How do I suppress vips warnings after this change?

**Comment by jcupitt** [comment]
8.5 now has a hack which will suppress warnings if the env var `VIPS_WARNING` is defined.

**Comment by felixbuenemann** [comment]
Will the change be included into 8.6/master as well? It seems to be very little overhead for keeping backwards compatibility.

**Comment by jcupitt** [comment]
Yes, 8.6 has it too.
